res/controllers/lib/midi-controls-0.js

```diff
@@ -0,0 +1,651 @@
+self-documenting.
+**/
+var RingEncoder = function (options) {
+    Control.call(this, options);
+};
+RingEncoder.prototype = new Control();
+
+var ControlContainer = function (initialLayer) {
+    if (typeof initialLayer === 'object') {
+        this.applyLayer(initialLayer);
+    }
+};
+ControlContainer.prototype = {
+    forEachControl: function (operation, recursive) {
+        if (typeof operation !== 'function') {
+            print('ERROR: ControlContainer.forEachContainer requires a function argument');
+            return;
+        }
+        if (recursive === undefined) { recursive = true; }
+
+        var that = this;
+        var applyOperationTo = function (obj) {
+            if (obj instanceof Control) {
+                operation.call(that, obj);
+            } else if (recursive && obj instanceof ControlContainer) {
+                obj.forEachControl(operation);
+            } else if (Array.isArray(obj)) {
+                obj.forEach(function (element) {
+                    applyOperationTo(element);
+                });
+            }
+        };
+
+        for (var memberName in this) {
+            if (this.hasOwnProperty(memberName)) {
+                applyOperationTo(this[memberName]);
+            }
+        }
+    },
+    reconnectControls: function (operation, recursive) {
+        this.forEachControl(function (control) {
+            control.disconnect();
+            if (typeof operation === 'function') {
+                operation.call(this, control);
+            }
+            control.connect();
+            control.trigger();
+        }, recursive);
+    },
+    isShifted: false,
+    shift: function () {
+        this.forEachControl(function (control) {
+            if (typeof control.shift === 'function') {
+                control.shift();
+            }
+            // Set isShifted for child ControlContainers forEachControl is iterating through recursively
+            this.isShifted = true;
+        });
+    },
+    unshift: function () {
+        this.forEachControl(function (control) {
+            if (typeof control.unshift === 'function') {
+                control.unshift();
+            }
+            // Set isShifted for child ControlContainers forEachControl is iterating through recursively
+            this.isShifted = false;
+        });
+    },
+    applyLayer: function (newLayer, reconnectControls) {
+        if (reconnectControls !== false) {
+            reconnectControls = true;
+        }
+        if (reconnectControls === true) {
+            this.forEachControl(function (control) {
+                control.disconnect();
+            });
+        }
+
+        _.merge(this, newLayer);
+
+        if (reconnectControls === true) {
+            this.forEachControl(function (control) {
+                control.connect();
+                control.trigger();
+            });
+        }
+    },
+};
+
+script.samplerRegEx = /\[Sampler(\d+)\]/ ;
```

Comment:
I don't think you should change other libraries from here :). Want to move these to the common scripts or make them a local variable?